File: ASP.NET Core Web API using PostgreSQL/Controllers/CollaborativeEditingController.cs
using Microsoft.AspNetCore.Mvc;
using Syncfusion.EJ2.DocumentEditor;
using Microsoft.AspNetCore.Cors;
using Microsoft.AspNetCore.SignalR;
using WebApplication1.Hubs;
using Microsoft.Data.SqlClient;
using System.Data;
using Microsoft.CodeAnalysis;
using Npgsql;
using Microsoft.EntityFrameworkCore.Metadata.Internal;
using static Microsoft.EntityFrameworkCore.DbLoggerCategory.Database;
using Newtonsoft.Json.Linq;

namespace WebApplication1.Controllers
{
    [Route("api/[controller]")]
    [ApiController]
    public class CollaborativeEditingController : ControllerBase
    {
        private readonly IWebHostEnvironment _hostingEnvironment;
        private readonly IHubContext<DocumentEditorHub> _hubContext;
        private static string connectionString;
        private static string fileLocation;
        private static byte saveThreshold = 200;

        public CollaborativeEditingController(IWebHostEnvironment hostingEnvironment, IHubContext<DocumentEditorHub> hubContext, IConfiguration config)
        {
            _hostingEnvironment = hostingEnvironment;
            _hubContext = hubContext;
            //Database connection string
            connectionString = config.GetConnectionString("DocumentEditorDatabase");
            fileLocation = _hostingEnvironment.WebRootPath;
        }

        //Import document from wwwroot folder in web server.
        [HttpPost]
        [Route("ImportFile")]
        [EnableCors("AllowAllOrigins")]
        public string ImportFile([FromBody] FileInfo param)
        {
            DocumentContent content = new DocumentContent();
            WordDocument document = GetSourceDocument(param.fileName);
            int lastSyncedVersion = 0;
            List<ActionInfo> actions = CreatedTable(param.roomName, out lastSyncedVersion);
            if (actions != null)
            {
                //Updated pending edit from database to source document.
                document.UpdateActions(actions);
            }
            string json = Newtonsoft.Json.JsonConvert.SerializeObject(document);
            content.version = lastSyncedVersion;
            content.sfdt = json;
            return Newtonsoft.Json.JsonConvert.SerializeObject(content);
        }

        [HttpPost]
        [Route("UpdateAction")]
        [EnableCors("AllowAllOrigins")]
        public async Task<ActionInfo> UpdateAction([FromBody] ActionInfo param)
        {
            try
            {
                ActionInfo modifiedAction = AddOperationsToTable(param);
                await _hubContext.Clients.Group(param.RoomName).SendAsync("dataReceived", "action", modifiedAction);
                return modifiedAction;
            }
            catch
            {
                return null;
            }
        }

        [HttpPost]
        [Route("GetActionsFromServer")]
        [EnableCors("AllowAllOrigins")]
        public string GetActionsFromServer([FromBody] ActionInfo param)
        {
            string tableName = param.RoomName;
            string getOperation = "SELECT * FROM \"" + tableName + "\" WHERE version > " + param.Version;
            using (NpgsqlConnection connection = new NpgsqlConnection(connectionString))
            {
                try
                {
                    NpgsqlCommand command2 = new NpgsqlCommand(getOperation, connection);
                    NpgsqlCommand updateCommand = new NpgsqlCommand(getOperation, connection);
                    connection.Open();
                    NpgsqlDataReader reader = updateCommand.ExecuteReader();
                    DataTable table = new DataTable();
                    table.Load(reader);
                    DataTable oldTable = table;
                    if (table.Rows.Count > 0)
                    {
                        int startVersion = int.Parse(table.Rows[0]["version"].ToString());
                        int lowestVersion = GetLowestClientVersion(table);
                        if (startVersion > lowestVersion)
                        {
                            string updatedOperation = "SELECT * FROM \"" + tableName + "\" WHERE version >= " + lowestVersion;
                            NpgsqlCommand command = new NpgsqlCommand(updatedOperation, connection);
                            NpgsqlDataReader reader2 = command.ExecuteReader();
                            table = new DataTable();
                            table.Load(reader2);
                        }
                        List<ActionInfo> actions = GetOperationsQueue(table);
                        foreach (ActionInfo info in actions)
                        {
                            if (!info.IsTransformed)
                            {
                                CollaborativeEditingHandler.TransformOperation(info, actions);
                            }
                        }
                        actions = actions.Where(x => x.Version > param.Version).ToList();
                        return Newtonsoft.Json.JsonConvert.SerializeObject(actions);
                    }
                }
                catch
                {
                    return "{}";
                }
            }
            return "{}";
        }

        private static WordDocument GetSourceDocument(string fileName)
        {
            string path = fileLocation + "\\" + fileName;
            int index = fileName.LastIndexOf('.');
            string type = index > -1 && index < fileName.Length - 1 ?
              fileName.Substring(index) : ".docx";
            Stream stream = System.IO.File.Open(path, FileMode.Open, FileAccess.Read, FileShare.Read);
            WordDocument document = Syncfusion.EJ2.DocumentEditor.WordDocument.Load(stream, GetFormatType(type));
            stream.Dispose();
            return document;
        }

        private List<ActionInfo> CreatedTable(string roomName, out int lastSyncedVersion)
        {
            lastSyncedVersion = 0;
            string tableName = roomName;
            if (!TableExists(tableName))
            {

                string queryString = "CREATE TABLE \"" + roomName
                   + "\" (version SERIAL PRIMARY KEY, operation TEXT, clientVersion INTEGER)";             
                using (NpgsqlConnection connection = new NpgsqlConnection(connectionString))
                {
                    NpgsqlCommand command = new NpgsqlCommand(queryString, connection);
                    connection.Open();
                    command.ExecuteNonQuery();
                    // Create table to track the last saved version.
                    CreateRecordForVersionInfo(connection, roomName);
                }
            }
            else
            {

                using (NpgsqlConnection connection = new NpgsqlConnection(connectionString))
                {

                    connection.Open();
                    string queryString = "SELECT * FROM \"" + tableName + "\" WHERE version > " + lastSyncedVersion;
                    NpgsqlCommand command = new NpgsqlCommand(queryString, connection);
                    connection.Open();                 
                    NpgsqlDataReader reader = command.ExecuteReader();
                    DataTable table = new DataTable();
                    table.Load(reader);
                    List<ActionInfo> actions = GetOperationsQueue(table);
                    return actions;

                }
            }
            return null;
        }
        private void CreateRecordForVersionInfo(NpgsqlConnection connection, String roomName)
        {
            string tableName = "de_version_info";         

            // Check if table exists
            if (!TableExists(tableName))
            {
                // If table doesn't exist, create it
                string createTableQuery = $"CREATE TABLE \"" + tableName + "\" (roomName TEXT, lastSavedVersion INTEGER)"; ;
                using (NpgsqlCommand createTableCommand = new NpgsqlCommand(createTableQuery, connection))
                {
                    createTableCommand.ExecuteNonQuery();
                }
            }

            // Insert record into the table
            string insertQuery = $"INSERT INTO \"" + tableName + "\" (roomName, lastSavedVersion) VALUES (?, ?)";
            using (NpgsqlCommand insertCommand = new NpgsqlCommand(insertQuery, connection))
            {
                insertCommand.Parameters.AddWithValue("@roomName", roomName);
                // Set initial version to 0
                insertCommand.Parameters.AddWithValue("@lastSavedVersion", 0);
                insertCommand.ExecuteNonQuery();
            }
            //}

        }
        private bool TableExists(string roomName)
        {
            using (var connection = new NpgsqlConnection(connectionString))
            {
                 NpgsqlCommand command = new NpgsqlCommand($"SELECT CASE WHEN EXISTS (SELECT 1 FROM information_schema.tables WHERE table_name = '"
                    + roomName + "') THEN 1 ELSE 0 END;", connection);
                connection.Open();
                var result = (int)command.ExecuteScalar();
                return result == 1;
            }
        }

        private ActionInfo AddOperationsToTable(ActionInfo action)
        {
            int clientVersion = action.Version;
            string tableName = action.RoomName;
            string value = Newtonsoft.Json.JsonConvert.SerializeObject(action);
            string query = $"INSERT INTO \"{tableName}\" (operation, clientVersion) VALUES (@Operation, @ClientVersion) RETURNING version AS last_id;";

            using (NpgsqlConnection connection = new NpgsqlConnection(connectionString))
            {

                NpgsqlCommand command = new NpgsqlCommand(query, connection);
                command.Parameters.AddWithValue("@Operation", value);
                command.Parameters.AddWithValue("@ClientVersion", action.Version);

                connection.Open();
                int updateVersion = (int)command.ExecuteScalar();
                if (updateVersion - clientVersion == 1)
                {
                    action.Version = updateVersion;
                    UpdateCurrentActionToDB(tableName, action, connection);
                }
                else
                {
                    DataTable table = GetOperationsToTransform(tableName, clientVersion + 1, updateVersion, connection);
                    int startVersion = int.Parse(table.Rows[0]["version"].ToString());
                    int lowestVersion = GetLowestClientVersion(table);
                    if (startVersion > lowestVersion)
                    {
                        table = GetOperationsToTransform(tableName, lowestVersion, updateVersion, connection);
                    }
                    List<ActionInfo> actions = GetOperationsQueue(table);
                    foreach (ActionInfo info in actions)
                    {
                        if (!info.IsTransformed)
                        {
                            CollaborativeEditingHandler.TransformOperation(info, actions);
                        }
                    }
                    action = actions[actions.Count - 1];
                    action.Version = updateVersion;
                    UpdateCurrentActionToDB(tableName, actions[actions.Count - 1], connection);
                }
                if (updateVersion % saveThreshold == 0)
                {
                    UpdateOperationsToSourceDocument(tableName, HttpContext.Session.GetString("UserId"), true, updateVersion);
                }


            }
            return action;
        }

        private void UpdateCurrentActionToDB(string tableName, ActionInfo action, NpgsqlConnection connection)
        {
            action.IsTransformed = true;
            string updateQuery = "UPDATE \"" + tableName + "\" SET operation = ? WHERE version = " + action.Version.ToString();
            NpgsqlCommand updateCommand = new NpgsqlCommand(updateQuery, connection);
            updateCommand.Parameters.AddWithValue("@Operation", Newtonsoft.Json.JsonConvert.SerializeObject(action));
            updateCommand.ExecuteNonQuery();
        }

        private static DataTable GetOperationsToTransform(string tableName, int clientVersion, int currentVersion, NpgsqlConnection connection)
        {
            string getOperation = "SELECT * FROM \"" + tableName + "\" WHERE version BETWEEN " + clientVersion + " AND " + currentVersion.ToString();
            NpgsqlCommand command = new NpgsqlCommand(getOperation, connection);
            NpgsqlDataReader reader = command.ExecuteReader();
            DataTable table = new DataTable();
            table.Load(reader);
            return table;
        }

        private static List<ActionInfo> GetOperationsQueue(DataTable table)
        {
            List<ActionInfo> actions = new List<ActionInfo>();
            foreach (DataRow row in table.Rows)
            {
                ActionInfo action = Newtonsoft.Json.JsonConvert.DeserializeObject<ActionInfo>(row["operation"].ToString());
                action.Version = int.Parse(row["version"].ToString());
                action.ClientVersion = int.Parse(row["clientVersion"].ToString());
                actions.Add(action);
            }
            return actions;
        }

        private static int GetLowestClientVersion(DataTable table)
        {
            int clientVersion = int.Parse(table.Rows[0]["clientVersion"].ToString());
            foreach (DataRow row in table.Rows)
            {
                //TODO: Need to optimise version calculation for only untransformed operations
                int version = int.Parse(row["clientVersion"].ToString());
                if (version < clientVersion)
                {
                    clientVersion = version;
                }
            }
            return clientVersion;
        }

        internal static FormatType GetFormatType(string format)
        {
            if (string.IsNullOrEmpty(format))
                throw new NotSupportedException("EJ2 DocumentEditor does not support this file format.");
            switch (format.ToLower())
            {
                case ".dotx":
                case ".docx":
                case ".docm":
                case ".dotm":
                    return FormatType.Docx;
                case ".dot":
                case ".doc":
                    return FormatType.Doc;
                case ".rtf":
                    return FormatType.Rtf;
                case ".txt":
                    return FormatType.Txt;
                case ".xml":
                    return FormatType.WordML;
                case ".html":
                    return FormatType.Html;
                default:
                    throw new NotSupportedException("EJ2 DocumentEditor does not support this file format.");
            }
        }

        /// <summary>
        /// Update editing operation to source document.
        /// </summary>
        public static void UpdateOperationsToSourceDocument(string fileName, string userId, bool partialSave, int endVersion)
        {
            try
            {
                NpgsqlConnection connection = new NpgsqlConnection(connectionString);
                connection.Open();
                string tableName = fileName;
                int lastSyncedVersion = GetLastedSyncedVersion(connection, fileName);
                string getOperation = "";
                if (partialSave)
                {
                    getOperation = "SELECT * FROM \"" + tableName + "\" WHERE version BETWEEN " + (lastSyncedVersion + 1).ToString() + " AND " + endVersion.ToString();                  
                }
                else
                {
                    getOperation = "SELECT * FROM \"" + tableName + "\" WHERE version > " + lastSyncedVersion;
                }
                NpgsqlCommand command = new NpgsqlCommand(getOperation, connection);
                NpgsqlDataReader reader = command.ExecuteReader();
                DataTable table = new DataTable();
                table.Load(reader);
                if (table.Rows.Count > 0)
                {
                    List<ActionInfo> actions = GetOperationsQueue(table);
                    foreach (ActionInfo info in actions)
                    {
                        if (!info.IsTransformed)
                        {
                            CollaborativeEditingHandler.TransformOperation(info, actions);
                        }
                    }
                    var currentDirectory = System.IO.Directory.GetCurrentDirectory();
                    int index = fileName.LastIndexOf('.');
                    string type = index > -1 && index < fileName.Length - 1 ?
                    fileName.Substring(index) : ".docx";
                    Stream stream1 = System.IO.File.Open(currentDirectory + "\\" + fileName, FileMode.Open, FileAccess.ReadWrite);
                    Syncfusion.EJ2.DocumentEditor.WordDocument document = Syncfusion.EJ2.DocumentEditor.WordDocument.Load(stream1, GetFormatType(type));
                    stream1.Close();
                    CollaborativeEditingHandler handler = new CollaborativeEditingHandler(document);
                    for (int i = 0; i < actions.Count; i++)
                    {
                        //Console.WriteLine(i);
                        handler.UpdateAction(actions[i]);
                    }
                    MemoryStream stream = new MemoryStream();
                    Syncfusion.DocIO.DLS.WordDocument doc = WordDocument.Save(Newtonsoft.Json.JsonConvert.SerializeObject(handler.Document));
                    doc.Save(stream, Syncfusion.DocIO.FormatType.Docx);
                    stream.Position = 0;
                    byte[] data = stream.ToArray();
                    System.IO.File.WriteAllBytes(currentDirectory + "\\output.docx", data);
                    stream.Close();
                    if (!partialSave)
                    {
                        endVersion = actions[actions.Count - 1].Version;
                    }
                    doc.Close();
                }
                if (!partialSave)
                {
                    DeleteLastModifiedVersion(tableName, connection);
                    DropTable(fileName, connection);

                }
                else
                {
                    UpdateModifiedVersion(tableName, connection, endVersion);

                }

            }
            catch (Exception ex)
            {

            }

        }
        static void UpdateModifiedVersion(string roomName, NpgsqlConnection connection, int lastSavedVersion)
        {
            string tableName = "de_version_info";
            string query = "UPDATE \"" + tableName + "\" SET lastSavedVersion = ? where roomName= '" + roomName + "'";

            using (NpgsqlCommand command = new NpgsqlCommand(query, connection))
            {
                command.Parameters.AddWithValue("@lastSavedVersion", lastSavedVersion);
                command.Parameters.AddWithValue("@roomName", roomName);
                command.ExecuteNonQuery();
            }
        }
        static void DeleteLastModifiedVersion(string roomName, NpgsqlConnection connection)
        {
            string tableName = "de_version_info";
            string query = "DELETE FROM \"" + tableName + "\" WHERE roomName= '" + roomName + "'";

            using (NpgsqlCommand command = new NpgsqlCommand(query, connection))
            {
                command.Parameters.AddWithValue("@roomName", roomName);
                command.ExecuteNonQuery();
            }
        }
        private static int GetLastedSyncedVersion(NpgsqlConnection connection, string roomName)
        {
            string tableName = "de_version_info";
            string query = "SELECT lastSavedVersion FROM \"" + tableName + "\" WHERE roomName ='" + roomName + "'";
            var command = new NpgsqlCommand(query, connection);
            command.Parameters.AddWithValue("@Operation", roomName);            
            return int.Parse(command.ExecuteScalar().ToString());
        }
        private static void DropTable(string documentId, NpgsqlConnection connection)
        {
            try
            {
                //Delete operations record.
                string sqlQuery = "drop table \"" + documentId + "\"";
                var sqlCommand = new NpgsqlCommand(sqlQuery, connection);
                sqlCommand.ExecuteNonQuery();
            }
            catch (Exception e)
            {
                Console.WriteLine(e.ToString());
            }
        }

    }


    public class FileInfo
    {
        public string fileName
        {
            get;
            set;
        }
        public string roomName
        {
            get;
            set;
        }

    }

    public class DocumentContent
    {
        public int version { get; set; }

        public string sfdt { get; set; }
    }
}

File: Server side with database/Using Database/Java/Java web service using PostgreSQL/src/main/java/com/syncfusion/tomcat/CollaborativeEditingController.java
package com.syncfusion.tomcat;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

import javax.sql.DataSource;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.messaging.MessageHeaders;
import org.springframework.web.bind.annotation.CrossOrigin;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RestController;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonMappingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.google.gson.Gson;
import com.syncfusion.docio.WordDocument;
import com.syncfusion.ej2.wordprocessor.ActionInfo;
import com.syncfusion.ej2.wordprocessor.CollaborativeEditingHandler;
import com.syncfusion.ej2.wordprocessor.WordProcessorHelper;
import com.syncfusion.javahelper.system.OutSupport;
import com.syncfusion.tomcat.controller.DocumentEditorHub;

import software.amazon.awssdk.auth.credentials.AwsBasicCredentials;
import software.amazon.awssdk.auth.credentials.AwsCredentials;
import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider;
import software.amazon.awssdk.core.ResponseInputStream;
import software.amazon.awssdk.regions.Region;
import software.amazon.awssdk.services.s3.S3Client;
import software.amazon.awssdk.services.s3.model.GetObjectRequest;
import software.amazon.awssdk.services.s3.model.GetObjectResponse;
import software.amazon.awssdk.services.s3.model.PutObjectRequest;

@RestController
public class CollaborativeEditingController {
	private static short saveThreshold = 200;
	// SQL
	@Value("${spring.datasource.url}")
	public String datasourceUrl;
	@Value("${spring.datasource.username}")
	private String datasourceUsername;
	@Value("${spring.datasource.password}")
	private String datasourcePassword;

	// Bucket S3
	@Value("${spring.datasource.accesskey}")
	private String datasourceAccessKey;
	@Value("${spring.datasource.secretkey}")
	private String datasourceSecretKey;
	@Value("${spring.datasource.bucketname}")
	private String datasourceBucketName;
	@Value("${spring.datasource.regionname}")
	private String datasourceRegionName;

	@Autowired
	private DataSource dataSource;

	@CrossOrigin(origins = "*", allowedHeaders = "*")
	@PostMapping("/api/collaborativeediting/ImportFile")
	public String ImportFile(@RequestBody FilesPathInfo file) throws Exception {
		try {
			DocumentContent content = new DocumentContent();

			// Load the document from S3 bucket.
			WordProcessorHelper document = getDocumentFromBucketS3(file.getFileName(), datasourceAccessKey,
					datasourceSecretKey, datasourceBucketName);

			int lastSyncedVersion = 0;
			OutSupport<Integer> lastSyncedVersion_out = new OutSupport<Integer>(Integer.class);
			// Create table in database to store temporary data for the collaborative
			// editing session.
			// Room name is a unique identifier of the document for collaborative editing.
			// We need to maintain the unique ID to map it with the original document to
			// save it in the S3.			
				
			ArrayList<ActionInfo> actions = createRecordForCollaborativeEditing(file.getRoomName(),
					lastSyncedVersion_out);

			if (actions != null && actions.size() > 0) {
				// When new user join the collaborative editing session
				// Get previous editing operation from DB and update it in document editor.
				document.updateActions(actions);
				ActionInfo lastAction = actions.get(actions.size() - 1);
				lastSyncedVersion = lastAction.getVersion();
			}
			String json = WordProcessorHelper.serialize(document);
			content.setVersion(lastSyncedVersion);
			content.setSfdt(json);
			Gson gson = new Gson();
			String data = gson.toJson(content);
			return data;
		} catch (Exception e) {
			e.printStackTrace();
			return "{\"sections\":[{\"blocks\":[{\"inlines\":[{\"text\":" + e.getMessage() + "}]}]}]}";
		}
	}

	@CrossOrigin(origins = "*", allowedHeaders = "*")
	@PostMapping("/api/collaborativeediting/UpdateAction")
	public ActionInfo UpdateAction(@RequestBody ActionInfo param) throws Exception {
		String roomName = param.getRoomName();
		ActionInfo transformedAction = addOperationsToTable(param);
		HashMap<String, Object> action = new HashMap<>();
		action.put("action", "updateAction");
		DocumentEditorHub.publishToRedis(roomName, transformedAction);
		//DocumentEditorHub.broadcastToRoom(roomName, transformedAction, new MessageHeaders(action));
		return transformedAction;
	}

	@CrossOrigin(origins = "*", allowedHeaders = "*")
	@PostMapping("/api/collaborativeediting/GetActionsFromServer")
	public String GetActionsFromServer(@RequestBody ActionInfo param) throws ClassNotFoundException {
		String tableName = param.getRoomName();
		String getOperation = "SELECT * FROM \"" + tableName + "\" WHERE version > " + param.getVersion();
		try (Connection connection = dataSource.getConnection()) {
			try (PreparedStatement updateCommand = connection.prepareStatement(getOperation)) {
				ResultSet reader = updateCommand.executeQuery();
				ArrayList<ActionInfo> actions;
				ArrayList<Map<String, Object>> table = getOperationsFromDatabaseResult(reader);
				if (!table.isEmpty()) {
					int startVersion = Integer.parseInt(table.get(0).get("version").toString());
					int lowestVersion = getLowestClientVersion(table);
					if (startVersion > lowestVersion) {
						String updatedOperation = "SELECT * FROM \"" + tableName + "\" WHERE version >= "
								+ lowestVersion;
						try (PreparedStatement command = connection.prepareStatement(updatedOperation)) {
							ResultSet reader2 = command.executeQuery();
							table.clear();
							table = getOperationsFromDatabaseResult(reader2);
						}
					}
					actions = getOperationsQueue(table);
					for (ActionInfo info : actions) {
						if (!info.isTransformed()) {
							CollaborativeEditingHandler.transformOperation(info, actions);
						}
					}
					// Assuming getIsTransformed() and Version are properties in ActionInfo
					actions.removeIf(x -> x.getVersion() <= param.getVersion());
					return new com.fasterxml.jackson.databind.ObjectMapper().writeValueAsString(actions);
				}
			}
		} catch (Exception e) {
			e.printStackTrace();
		}
		return "{}";
	}

	private ActionInfo addOperationsToTable(ActionInfo action) throws Exception {
		int clientVersion = action.getVersion();
		String roomName = action.getRoomName();
		String value = new com.fasterxml.jackson.databind.ObjectMapper().writeValueAsString(action);
		String query = "INSERT INTO \"" + roomName + "\" (operation, clientVersion) VALUES (?, ?) RETURNING version;";

		try (Connection connection = dataSource.getConnection()) {
			PreparedStatement preparedStatement = connection.prepareStatement(query);
			preparedStatement.setString(1, value);
			preparedStatement.setInt(2, action.getVersion());
			try {
				ResultSet resultSet = preparedStatement.executeQuery();
				if (resultSet.next()) {
					int updateVersion = resultSet.getInt("version");
					if (updateVersion - clientVersion == 1) {
						action.setVersion(updateVersion);
						updateCurrentActionToDB(roomName, action, connection);
					} else {
						ArrayList<Map<String, Object>> table = getOperationsToTransform(roomName, clientVersion + 1,
								updateVersion, connection);
						int startVersion = (int) table.get(0).get("version");
						int lowestVersion = getLowestClientVersion(table);
						if (startVersion > lowestVersion) {
							table = getOperationsToTransform(roomName, lowestVersion, updateVersion, connection);
						}
						ArrayList<ActionInfo> actions = getOperationsQueue(table);
						for (ActionInfo info : actions) {
							if (!info.isTransformed()) {
								CollaborativeEditingHandler.transformOperation(info, actions);
							}
						}
						action = actions.get(actions.size() - 1);
						action.setVersion(updateVersion);
						updateCurrentActionToDB(roomName, actions.get(actions.size() - 1), connection);
					}
					if (updateVersion % saveThreshold == 0) {
						// Update the operations to table once specified threshold is reached.
						updateOperationsToSourceDocument(roomName, true, updateVersion, connection, datasourceAccessKey,
								datasourceSecretKey, datasourceBucketName);
					}
				}
			} catch (Exception e) {
				e.printStackTrace();
			}
		}
		return action;
	}

	private int getLowestClientVersion(ArrayList<Map<String, Object>> table) {
		int clientVersion = (int) table.get(0).get("clientVersion");
		for (Map<String, Object> row : table) {
			int version = Integer.parseInt(row.get("clientVersion").toString());
			if (version < clientVersion) {
				clientVersion = version;
			}
		}
		return clientVersion;
	}

	private ArrayList<Map<String, Object>> getOperationsToTransform(String tableName, int clientVersion,
			int currentVersion, Connection connection) throws ClassNotFoundException {
		String getOperation = "SELECT * FROM \"" + tableName + "\" WHERE version BETWEEN ? AND ?";
		try (PreparedStatement command = connection.prepareStatement(getOperation)) {
			command.setInt(1, clientVersion);
			command.setInt(2, currentVersion);
			try (ResultSet resultSet = command.executeQuery()) {
				ArrayList<Map<String, Object>> resultList = getOperationsFromDatabaseResult(resultSet);
				return resultList;
			}
		} catch (SQLException e) {
			e.printStackTrace();
			return null;
		}
	}

	private void updateCurrentActionToDB(String tableName, ActionInfo action, Connection connection) {
		action.setTransformed(true);
		String updateQuery = "UPDATE \"" + tableName + "\" SET operation = ? WHERE version = " + action.getVersion();
		try (PreparedStatement updateStatement = connection.prepareStatement(updateQuery)) {
			updateStatement.setString(1, new com.fasterxml.jackson.databind.ObjectMapper().writeValueAsString(action));
			updateStatement.executeUpdate();
		} catch (SQLException | com.fasterxml.jackson.core.JsonProcessingException e) {
			e.printStackTrace();
		}
	}

	public static void updateOperationsToSourceDocument(String roomName, boolean partialSave, int endVersion,
			Connection connection, String accessKey, String secretKey, String bucketName) throws Exception {

		int lastSyncedVersion = getLastSavedVersion(connection, roomName);
		String getOperation = "";
		if (partialSave) {
			getOperation = "SELECT * FROM \"" + roomName + "\" WHERE version BETWEEN " + (lastSyncedVersion + 1)
					+ " AND " + endVersion;
		} else {
			getOperation = "SELECT * FROM \"" + roomName + "\" WHERE version > " + lastSyncedVersion;
		}

		try (PreparedStatement command = connection.prepareStatement(getOperation)) {
			ResultSet reader = command.executeQuery();
			ArrayList<Map<String, Object>> table = getOperationsFromDatabaseResult(reader);
			ArrayList<ActionInfo> actions = getOperationsQueue(table);
			for (ActionInfo info : actions) {
				if (!info.isTransformed()) {
					CollaborativeEditingHandler.transformOperation(info, actions);
				}
			}
			reader.close();
			command.close();

			// room name is unique identifier used for editing the document in collaborative
			// editing.
			// Please get the actual file name by mapping the room name.
			String fileName = "";

			CollaborativeEditingHandler handler = new CollaborativeEditingHandler(
					getDocumentFromBucketS3(fileName, accessKey, secretKey, bucketName));
			for (ActionInfo info : actions) {
				handler.updateAction(info);
			}
			ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
			WordDocument doc = WordProcessorHelper.save(WordProcessorHelper.serialize(handler.getDocument()));
			doc.save(outputStream, com.syncfusion.docio.FormatType.Docx);
			byte[] data = outputStream.toByteArray();
			outputStream.close();
			doc.close();

			// Save the document to S3 bucket
			AwsCredentials credentials = AwsBasicCredentials.create(accessKey, secretKey);
			StaticCredentialsProvider credentialsProvider = StaticCredentialsProvider.create(credentials);
			S3Client s3Client = S3Client.builder().region(Region.US_EAST_1).credentialsProvider(credentialsProvider)
					.build();
			PutObjectRequest objectRequest = PutObjectRequest.builder().bucket(bucketName).key(roomName).build();
			s3Client.putObject(objectRequest, software.amazon.awssdk.core.sync.RequestBody.fromBytes(data));
			s3Client.close();

			if (!partialSave) {
				deleteLastModifiedVersion(roomName, connection);
				dropTable(roomName, connection);
			} else {
				updateModifiedVersion(roomName, connection, endVersion);
			}
		}

		// connection.commit();
		/*
		 * } catch (Exception e) { e.printStackTrace(); }
		 */
	}

	private static void updateModifiedVersion(String roomName, Connection connection, int lastSavedVersion)
			throws SQLException {
		String tableName = "de_version_info";
		String query = "UPDATE \"" + tableName + "\" SET lastSavedVersion = ? where roomName= '" + roomName + "'";
		PreparedStatement preparedStatement = connection.prepareStatement(query);
		preparedStatement.setInt(1, lastSavedVersion);
		preparedStatement.executeUpdate();
	}

	private static void deleteLastModifiedVersion(String roomName, Connection connection) throws SQLException {
		String tableName = "de_version_info";
		String query = "DELETE FROM \"" + tableName + "\" WHERE roomName= '" + roomName + "'";
		PreparedStatement preparedStatement = connection.prepareStatement(query);
		preparedStatement.executeUpdate();
	}

	public static void dropTable(String documentId, Connection connection) throws ClassNotFoundException {
		try {
			// Drop the table
			String sqlQuery = "DROP TABLE \"" + documentId + "\"";
			try (PreparedStatement sqlCommand = connection.prepareStatement(sqlQuery)) {
				sqlCommand.executeUpdate();
			}
		} catch (SQLException e) {
			e.printStackTrace();
		}
	}

	private ArrayList<ActionInfo> createRecordForCollaborativeEditing(String roomName,
			OutSupport<Integer> lastSyncedVersion) throws SQLException, ClassNotFoundException, IOException {
		try (Connection connection = dataSource.getConnection()) {
			if (!tableExists(connection, roomName)) {
				lastSyncedVersion.setValue(0);
				// If record not present, create new table
				String queryString = "CREATE TABLE \"" + roomName
						+ "\" (version SERIAL PRIMARY KEY, operation TEXT, clientVersion INTEGER)";
				PreparedStatement preparedStatement = connection.prepareStatement(queryString);
				preparedStatement.executeUpdate();
				// Create table to track the last saved version.
				createRecordForVersionInfo(connection, roomName);
			} else {
				lastSyncedVersion.setValue(getLastSavedVersion(connection, roomName));
				// If record present, get previous editing operation from the db record.
				String queryString = "SELECT * FROM \"" + roomName + "\" WHERE version > "
						+ lastSyncedVersion.getValue();
				try (PreparedStatement updateCommand = connection.prepareStatement(queryString)) {
					ResultSet result = updateCommand.executeQuery();
					ArrayList<ActionInfo> actions = getOperationsQueue(getOperationsFromDatabaseResult(result));
					return actions;
				}
			}
			connection.close();
		} catch (Exception e) {
			e.printStackTrace();
		}
		return null;
	}

	private static int getLastSavedVersion(Connection connection, String roomName) throws SQLException {
		String tableName = "de_version_info";
		String query = "SELECT lastSavedVersion FROM \"" + tableName + "\" WHERE roomName ='" + roomName + "'";
		PreparedStatement statement = connection.prepareStatement(query);
		ResultSet resultSet = statement.executeQuery();
		int lastSavedVersion = 0;
		if (resultSet.next()) {
			lastSavedVersion = resultSet.getInt("lastSavedVersion");
		}
		resultSet.close();
		statement.close();
		return lastSavedVersion;
	}

	private static void createRecordForVersionInfo(Connection connection, String roomName) throws SQLException {
		String tableName = "de_version_info";
		try {
			if (!tableExists(connection, tableName)) {
				// If record not present, create new table
				String queryString = "CREATE TABLE \"" + tableName + "\" (roomName TEXT, lastSavedVersion INTEGER)";
				PreparedStatement preparedStatement = connection.prepareStatement(queryString);
				preparedStatement.executeUpdate();
			}
			String query = "INSERT INTO \"" + tableName + "\" (roomName, lastSavedVersion) VALUES (?, ?)";
			PreparedStatement preparedStatement = connection.prepareStatement(query);
			preparedStatement.setString(1, roomName);
			// Set initial version to 0;
			preparedStatement.setInt(2, 0);
			preparedStatement.executeUpdate();
		} catch (ClassNotFoundException e) {
			e.printStackTrace();
		}
	}

	private static boolean tableExists(Connection connection, String roomName) throws ClassNotFoundException {
		try {
			String query = "SELECT CASE WHEN EXISTS (SELECT 1 FROM information_schema.tables WHERE table_name = '"
					+ roomName + "') THEN 1 ELSE 0 END;";
			PreparedStatement statement = connection.prepareStatement(query);
			// statement.setString(1, tableName);
			ResultSet resultSet = statement.executeQuery();
			if (resultSet.next()) {
				int result = resultSet.getInt(1);
				return result == 1;
			}
		} catch (Exception e) {
			e.printStackTrace();
		}
		return false;
	}

	private static ArrayList<Map<String, Object>> getOperationsFromDatabaseResult(ResultSet reader)
			throws SQLException {
		ArrayList<Map<String, Object>> table = new ArrayList<>();
		while (reader.next()) {
			Map<String, Object> row = new HashMap<>();
			row.put("version", reader.getInt("version"));
			row.put("clientVersion", reader.getInt("clientVersion"));
			row.put("operation", reader.getString("operation")); // Add other columns as needed
			table.add(row);
		}
		return table;
	}

	private static ArrayList<ActionInfo> getOperationsQueue(ArrayList<Map<String, Object>> table)
			throws JsonMappingException, JsonProcessingException {
		ArrayList<ActionInfo> actions = new ArrayList<>();
		for (Map<String, Object> row : table) {
			ObjectMapper objectMapper = new ObjectMapper();
			String jsonString = (String) row.get("operation");
			// Deserialize JSON string to ActionInfo object
			ActionInfo action = objectMapper.readValue(jsonString, ActionInfo.class);
			action.setVersion(Integer.parseInt(row.get("version").toString()));
			action.setClientVersion(Integer.parseInt(row.get("clientVersion").toString()));
			actions.add(action);
		}
		return actions;
	}

	private static WordProcessorHelper getDocumentFromBucketS3(String documentId, String accessKey, String secretKey,
			String bucketName) {	
		try {
			AwsCredentials credentials = AwsBasicCredentials.create(accessKey, secretKey);
			StaticCredentialsProvider credentialsProvider = StaticCredentialsProvider.create(credentials);
			S3Client s3Client = S3Client.builder().region(Region.US_EAST_1).credentialsProvider(credentialsProvider)
					.build();
			ResponseInputStream<GetObjectResponse> objectData = s3Client
					.getObject(GetObjectRequest.builder().bucket(bucketName).key(documentId).build());
			// Read the object data into a byte array
			ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
			byte[] buffer = new byte[1024];
			int bytesRead;
			while ((bytesRead = objectData.read(buffer)) != -1) {
				byteArrayOutputStream.write(buffer, 0, bytesRead);
			}
			s3Client.close();
			byte[] data = byteArrayOutputStream.toByteArray();
			// Create an input stream from the byte array
			try (InputStream stream = new ByteArrayInputStream(data)) {
				return WordProcessorHelper.load(stream, true);
			}
		} catch (Exception e) {
			e.printStackTrace();
		}
		return null;
	}
}
